### PR TITLE
HPX backend

### PR DESCRIPTION
**Pull request by msimberg**
This adds an HPX backend to the `Kokkos::Experimental` namespace. It contains (AFAICT) all functionality required of a Kokkos backend. It only runs with deprecated code disabled.

There is an HPX option called `enable_async_dispatch` which makes parallel dispatch asynchronous with the HPX backend. It's off by default. All work submitted through Kokkos is still processed sequentially. (By the way, I've gone and added a bunch of `Kokkos::fence` calls to tests to deal with the asynchronous version. I may have added them in too many places.)

A preprocessor option (`KOKKOS_HPX_IMPLEMENTATION`) switches between two different implementations for most of the parallel dispatch functions. One is the obvious HPX way of implementing the backend, but since HPX still lacks some optimizations there is a slightly more verbose and optimized version. The optimized version is the default. I haven't made a build system option for this since I don't expect any user to want to change that---it's mainly for testing.

I've tried to use existing Kokkos utilities as much as possible but some things I ended up implementing just for the HPX backend.

I've had it running within Trilinos (with HPX as a TPL) but I'm not very confident that I've done everything the correct tribits way. Comments on this would be appreciated.

KokkosKernels PR (and maybe Trilinos) will follow. I see there's no support for ROCm in KokkosKernels or Trilinos. Is it because you don't want to do that until it's out of the `Kokkos::Experimental` namespace or just because it hasn't been done yet?

Finally, let me know how you prefer to deal with:

- formatting: I've run `clang-format` with default settings on new files related to the HPX backend, and haven't changed formatting in other files. Is that OK?
- rebasing/squashing: I haven't cleaned up the history yet at all. Do you prefer one big squashed commit, or a few cleaned up commits?
- documentation: Is the wiki the place to document details about using the backend? E.g. I would recommend users to start the HPX runtime instead of letting Kokkos do it for performance reasons. Not that I expect people to flock to the HPX backend.

Edit: This PR also changes the calculation of chunk sizes to suit HPX better, but that can be separated out (especially if something like #1866 would be available).

### CODE REVIEW

**Comment by msimberg** [comment]
@mhoemmen Agreed. I've started that work. Tests pass in Kokkos and KokkosKernels, but not all tests pass in Tetra. Some also fail with the serial backend, and some seem to be unique to the HPX backend, but I haven't yet investigated whether these are caused by the HPX backend or HPX itself (or something else).

**Comment by crtrott** [comment]
Running SpotCheck now.

File: core/src/Kokkos_ExecPolicy.hpp [hunk 1]
```diff
@@ -220,11 +220,11 @@ class RangePolicy
    }
 
    member_type new_chunk_size = 1;
-   while(new_chunk_size*100*concurrency < m_end-m_begin)
```
Comment (crtrott) [inline core/src/Kokkos_ExecPolicy.hpp]: this is something we need to check. I am not sure that the heuristic to go this low is good on most platforms.

File: core/unit_test/TestAtomicViews.hpp [hunk 2]
```diff
@@ -112,6 +112,7 @@ struct TestViewOperator_LeftAndRight< DataType, DeviceType, 1 >
     int error_flag = 0;
 
     Kokkos::parallel_reduce( 1, driver, error_flag );
+    Kokkos::fence();
```
Comment (crtrott) [inline core/unit_test/TestAtomicViews.hpp]: Why do we need this? I guess HPX parallel_reduce is always asynchronous? The (unfortunately not fully yet written down) semantics of parallel_reduce are that it will fence if the result argument is a scalar. Only if you give it a view will it be asynchronous.

File: core/unit_test/TestMDRange.hpp [hunk 3]
```diff
@@ -2187,6 +2264,7 @@ struct TestMDRange_5D {
 
       TestMDRange_5D::HostViewType h_view = Kokkos::create_mirror_view( v );
       Kokkos::deep_copy( h_view, v );
+      Kokkos::fence();
```
Comment (crtrott) [inline core/unit_test/TestMDRange.hpp]: deep_copy without a execution space instance argument is synchronous with respect to the calling thread. deep_copy(HPX(),h_view,v); would be asynchronous with respect to the calling thread (but will fence against any other kernel launched on HPX).

**Comment by crtrott** [comment]
The tpetra issue might be because of the asynchronous nature of parallel_reduce and deep_copy in the default use case which actually should have synchronous semantics.

```c++
int result;
parallel_reduce(N,Functor,result); // Synchronous
Kokkos::View<int> result_view("R");
parallel_reduce(N,Functor,result_view); // Asynchronous

deep_copy(a,b); // Synchronous
deep_copy(exec_space_instance,a,b); // Asynchronous (Synchronous to other operations on exec_space_instance
```

File: core/src/Kokkos_ExecPolicy.hpp (hunk 1 above)
Comment (msimberg) [reply core/src/Kokkos_ExecPolicy.hpp]: Sure, that's understandable. I'm happy to remove this if it's fairly likely that I'll be able to customize this for the HPX backend at a later point.

File: core/unit_test/TestMDRange.hpp (hunk 3 above)
Comment (msimberg) [reply core/unit_test/TestMDRange.hpp]: Thanks for pointing that out. I'll remove those fences.

**Comment by msimberg** [comment]
You're right that the asynchronous `parallel_reduce` likely *will* be a problem in Tpetra, but so far I've only tested it with the synchronous version. I'll get back to you as soon as I've had time to look into the failures.

**Comment by msimberg** [comment]
Is [this](https://github.com/kokkos/kokkos/blob/master/core/src/Cuda/Kokkos_Cuda_Parallel.hpp#L960) what's implicitly leading to a fence with views, or is there something more/else to it?

**Comment by msimberg** [comment]
All right, thanks. That leaves me with one more question. `ParallelReduceReturnValue` wraps scalars in `Kokkos::View`s, even scalars, meaning that I only see `View`s and reducer objects in the backend. Should I really be checking for unmananaged `View`s, which is what scalars get wrapped in? I assume unmanaged `View`s explicitly passed to `parallel_reduce` should also block like scalars? What would be the recommended way to check for this? Is this accessible through `memory_traits` or something like that?

**Comment by crtrott** [comment]
@msimberg: That is an interesting question and somewhat underspecified right now. Maybe we should have a flag in the reducer which says whether it was constructed by capturing a scalar or a view. For now it may be safer to fence on unmanaged views. In Kokkos we don't have an asynchronous host backend right now hence this question never fully appeared, since the decision whether to be asynchronous was only posed for CUDA and only if we got a CudaSpace view for reducer. I think I like the idea of having a flag in the reducer to say whether it references a pure scalar. What do you think?

**Comment by msimberg** [comment]
Okay. Let me try to understand first why reduce with scalars should block, and views not. I thought this would be related to reference counting first, but that doesn't seem to be the case (a (managed) View could keep the data alive long enough for the reduction, but that get's disabled at tactical points in the code). Am I just misreading the code?

It seems like checking for unmanaged views is the more general solution (which would include user-passed unmanaged views), but I'm not sure I fully understand the reasons for implicitly blocking yet. Is this more about not allowing the user to shoot themselves in the foot or is there another technical reason for this behaviour. By the way, scan doesn't seem to impose this limitation. Am I allowed to pass a stack-allocated array (i.e. not a view) as the return value?

As for a fix, do you mean by reducer this: https://github.com/kokkos/kokkos/blob/5d6e7fb38e96aec88d2c514e1f9be1cf2b549b57/core/src/Kokkos_Parallel_Reduce.hpp#L743-L819? Or the actual reducer structs above that? Because

> if you give it a reducer ... it doesn't [block]

Adding a flag to either `struct ParallelReduceReturnValue` or the reducer structs would certainly be straightforward.

**Comment by crtrott** [comment]
I am talking about the actual reducer struct (Kokkos::Max() etc.). And the main reason for blocking is to making it harder for users to shoot themselves in the foot ;-). There is no technical reason really. Also the managed view thing to solve this problem for real is a thorny issue. Right now there is actually no good solution since even if a GPU kernel for example could reduce the reference count of all its views it can't call the dealloc. We would need to insert a cleanup host task into the cuda stream or something like that which would add significant latency overhead, which would be rather inconvenient for a lot of our codes (many of which are latency limited on how many kernels per second we can execute). If we were to return futures, we could keep the reference counts alive in there, and the destruction of the future would a) wait on the kernel to finish and b) decrease reference count and deallocate views if needed. 

Only allowing asynchronous execution with managed views might still be a good answer for now though. Managed views fence all of Kokkos at their deallocation in order to ensure that no kernels are in flight. If we do this we would have following situation:

```c++
int sum;
View<double> vsum("V");
View<double> vsum_um(&sum);
parallel_reduce("ParRedScalar",N,functor,sum); // blocking
parallel_reduce("ParRedScalar",N,functor,vsum); // non-blocking
parallel_reduce("ParRedScalar",N,functor,vsum_um); // blocking
parallel_reduce("ParRedScalar",N,functor,Sum<int>(sum)); // blocking
parallel_reduce("ParRedScalar",N,functor,Sum<int>(vsum)); // non-blocking
parallel_reduce("ParRedScalar",N,functor,Sum<int>(vsum_um)); // blocking
``` 

This sounds like a decent solution. What do you think?

**Comment by crtrott** [comment]
We also need to sort out the merge conflicts with the tasking update. Me or David will do this @dhollman

**Comment by msimberg** [comment]
@crtrott that looks like sane behaviour to me. `view.impl_track().has_record()` works nicely. I've updated `Kokkos::Sum` to do what I think you meant. Essentially a managed view constructed from an unmanaged view will be compile-time managed but runtime unmanaged, like below?

```
Kokkos::View<int, space, Unmanaged> unmanaged;
Kokkso::View<int, space> still_unmanaged(unmanaged);
```

If that looks reasonable I'll update the rest of the reducers to do the same. 

I've also hopefully removed all unnecessary fences after parallel reductions. Let me know if you think I've missed some. I have a fence after `Kokkos::wait` at the moment, because a task graph only starts executing in `wait`. This is admittedly quite ugly and I can make `wait` fence in the HPX backend if you prefer? I'm wondering if this behaviour will change? I would expect `wait` to actually be `fence` and something else to trigger execution before that.

**Comment by msimberg** [comment]
Ping @crtrott, @jeffmiles63?

**Comment by crtrott** [comment]
Sorry Michael, most of us are at the C++ committee meeting, and were busy preparing for it. We'll get this in next week. Promise :-)

**Comment by msimberg** [comment]
> Sorry Michael, most of us are at the C++ committee meeting, and were busy preparing for it. We'll get this in next week. Promise :-)

Ah, silly me, of course you are... Take your time, just wanted to make sure you're not waiting for me to do something at the moment.

`BLOCKS PROMOTION` looks scary. I expect you'll explain what that means once you get the time.

**Comment by crtrott** [comment]
BLOCKS_PROMOTION means this must be resolved (i.e. merged) before we do the next release.

**Comment by jeffmiles63** [comment]
Trying to unit test this against the current develop branch, and there appears to be issues with <TaskQueueSpecialization<Kokkos::Experimental::HPX> compiling. @msimberg ,  Could you please verify your implementation for TaskQueueSpecialization?

**Comment by msimberg** [comment]
@jeffmiles63 indeed, that broke after rebasing on latest master because of the changes to tasking. I'll have a look at that next week (if there's anything special to take into account there, please let me know @crtrott / @dhollman).

What's your (desired) timeline for the next release?

**Comment by crtrott** [comment]
Next release is gonna come probably early April. David is with me at the C++ committee I think he can fix that then if you don't find time. ;-)

**Comment by mhoemmen** [comment]
@msimberg FYI, as far as I know, best practice would be to rebase against latest develop, since that moves faster than master.

**Comment by msimberg** [comment]
> @msimberg FYI, as far as I know, best practice would be to rebase against latest develop, since that moves faster than master.

That's what I meant to say ;)

**Comment by msimberg** [comment]
Rebased and updated to work with the new tasking.

So the only thing missing would be dealing with the asynchronous parallel reductions and asynchronous task graph launch. Do you have any preferences on those?

**Comment by msimberg** [comment]
Any remaining concerns on this?

**Comment by dhollman** [comment]
FWIW, here's the spot-check of the rebased (+added `}` in that one place) code, even though spot-check obviously doesn't test the HPX backend itself:

```
Running on machine: apollo
Repository Status:  8fe65ad70e66a5b7a9b233778fde76ef4e6fd3bc rebased and added }


Going to test compilers:  gcc/4.8.4 gcc/5.3.0 intel/16.0.1 clang/3.9.0 clang/6.0 cuda/9.1
Testing compiler gcc/4.8.4
  Starting job gcc-4.8.4-OpenMP-release
  PASSED gcc-4.8.4-OpenMP-release
Testing compiler gcc/5.3.0
  Starting job gcc-4.8.4-Pthread-release
  PASSED gcc-4.8.4-Pthread-release
Testing compiler intel/16.0.1
  Starting job gcc-5.3.0-Serial-release
  PASSED gcc-5.3.0-Serial-release
Testing compiler clang/3.9.0
  Starting job intel-16.0.1-OpenMP-release
  PASSED intel-16.0.1-OpenMP-release
Testing compiler clang/6.0
  Starting job clang-3.9.0-Pthread_Serial-release
  PASSED clang-3.9.0-Pthread_Serial-release
  Starting job clang-6.0-Cuda_Pthread-release
  PASSED clang-6.0-Cuda_Pthread-release
Testing compiler cuda/9.1
  Starting job clang-6.0-OpenMP-release
  PASSED clang-6.0-OpenMP-release
  Starting job cuda-9.1-Cuda_OpenMP-release
  PASSED cuda-9.1-Cuda_OpenMP-release
#######################################################
PASSED TESTS
#######################################################
clang-3.9.0-Pthread_Serial-release build_time=137 run_time=474
clang-6.0-Cuda_Pthread-release build_time=236 run_time=330
clang-6.0-OpenMP-release build_time=117 run_time=94
cuda-9.1-Cuda_OpenMP-release build_time=319 run_time=184
gcc-4.8.4-OpenMP-release build_time=106 run_time=119
gcc-4.8.4-Pthread-release build_time=98 run_time=361
gcc-5.3.0-Serial-release build_time=114 run_time=181
intel-16.0.1-OpenMP-release build_time=334 run_time=155
```

File: core/src/Kokkos_HPX.hpp [hunk 4]
```diff
@@ -0,0 +1,1980 @@
...
+  const ReducerType m_reducer;
+  const pointer_type m_result_ptr;
+
+  bool m_force_synchronous;
+
+public:
+  void execute() const {
+    dispatch_execute_task(this);
+    if (m_force_synchronous) {
+      Experimental::HPX::fence();
```
[the remark this answers is not in the document]
Comment (msimberg) [reply core/src/Kokkos_HPX.hpp]: Thanks for noticing. Fixed.

**Comment by ibaned** [comment]
So... what is missing to merge this?

**Comment by dhollman** [comment]
> So... what is missing to merge this?

Some equivalent of `spot-check` that actually uses HPX?  The problem is that that would require compiling `HPX` on, for instance, `apollo`, which in turn requires compiling `boost` (or loading a module for boost) with MPI, which is not a trivial activity.

**Comment by ibaned** [comment]
@msimberg are you able to retest this with HPX in its current state?

**Comment by dhollman** [comment]
@ibaned more than that, though, to prevent code rot we should probably add some kind of testing of this to our nightlies or something like that.  I don't really know how we'd go about that (or if we should), but otherwise things will go wrong with this if we don't test new functionality on this backend.

**Comment by msimberg** [comment]
+1 on needing some regular testing on this. I'd be willing to run some kind of nightly testing myself to at least avoid it staying broken for a long time if something changes (and I'd be responsible for fixing it as well). That would be good enough for our purposes where we're just experimenting with using it as a backend for some applications. It's not a long term solution though.

This passes all the tests after the rebase as well except for an occasional deadlock which is caused by a bug in HPX itself (and I don't know how to fix it yet). I don't think that should stand in the way of merging this though.

**Comment by ndellingwood** [comment]
> This passes all the tests after the rebase as well except for an occasional deadlock which is caused by a bug in HPX itself (and I don't know how to fix it yet). I don't think that should stand in the way of merging this though.

@ibaned @crtrott @dhollman it sounds like this is ready to merge then?



>  I'd be willing to run some kind of nightly testing myself to at least avoid it staying broken for a long time if something changes (and I'd be responsible for fixing it as well).

@msimberg this would be great! We can also work on getting additional nightly testing as well, though this won't be ready in the near-term.

**Comment by msimberg** [comment]
FYI, this is going to do the nightly testing: https://github.com/msimberg/kokkos-hpx-ci/. The tests will of course only run after merges, so I'll also try to keep an eye out for PRs that might break the HPX backend.

**Comment by crtrott** [comment]
I am fixing some things with the HPX build using hex 1.2.1 for example it seems it needs to be lib64 not lib in the Makefile.kokkos for finding the pkg-config stuff.

**Comment by crtrott** [comment]
I am also adding an hpx build to Apollo.

**Comment by msimberg** [comment]
@crtrott thanks for the fixes! And great if it'll be tested more regularly by you guys as well (I'm assuming that's roughly what a build on Apollo means).

The `lib` to `lib64` change is obviously only for 64-bit systems, so really it should be checking for both `lib` and `lib64`. I'll push a fix for that in a moment.

I'm slightly worried that you'll be hitting the occasional timeouts I mentioned above. It's been fixed in master (https://github.com/STEllAR-GROUP/hpx/pull/3749) and we're aiming to have the next release out in May latest.

**Comment by dhollman** [comment]
This looks ready to merge from my perspective, now that it's passing on the CI that @msimberg provided.

**Comment by msimberg** [comment]
@dhollman it's not being tested yet! It's set up to test `develop` once a day, and since this isn't merged yet it's not tested. I'll change that to do this branch until it's merged.

**Comment by msimberg** [comment]
I just added a file I'd missed for the `feint` example. Current CI build is [here](https://circleci.com/gh/msimberg/kokkos-hpx-ci/17), but I'll let you know when I get it to pass.

**Comment by msimberg** [comment]
The synchronous backend is all good, but I hit a bug with the asynchronous backend which only appeared once I ran in the CI environment.

The backstory is that the CI image I used uses the system allocator, whereas I usually use `tcmalloc` for development. Tests and examples passed happily with `tcmalloc` but with the system allocator I had segfaults in certain examples and tests (for example https://github.com/kokkos/kokkos/blob/master/example/tutorial/Advanced_Views/03_subviews/subviews.cpp), presumably because of `tcmalloc` reusing or not freeing memory immediately.

Long story short, subviews don't increment the reference count, correct? (at least https://github.com/kokkos/kokkos/wiki/Subviews#subviews says so and is consistent with some of the segfaults I've seen).

**Comment by msimberg** [comment]
https://circleci.com/workflow-run/b9a9992e-23f5-44b8-a35f-53789f1f1eb4

**Comment by msimberg** [comment]
@crtrott you might want to check the fences in the second to last commit. They're necessary but their exact placement could be argued about.

Some of them could be avoided if View tracking could be disabled (in the parallel dispatch) per backend, rather than for all backends. That may cause all kinds of other problems that I'm not aware of though.

**Comment by crtrott** [comment]
Working on removing fences which shouldn't be necessary in the unit tests. The discussion we ended up with is that if you give parallel_reduce a view as place to store its value (managed, or unmanaged independent of memory space) directly or through a reducer it will be asynchronous. If the user provides a scalar (also a scalar wrapped into a reducer) the parallel_reduce call itself will fence. 

I have an idea for a Kokkos tool which can help identify places in an application where fences are potentially missing.

**Comment by crtrott** [comment]
See #2077

**Comment by crtrott** [comment]
Btw. @msimberg the issue above is the primary reason I didn't merge this yet. I am trying to come up with a way that the HPX backend a) works for our apps without having to add fences all over the place, and b) get our semantics nailed down. I am super grateful for your work on this - the HPX backend forces us to plug gaping holes in our specification - which were simply not an issue before. I am actually planning of making HPX backend pass builds a hard requirement for any future pull request being accepted.

**Comment by crtrott** [comment]
Running spotcheck now.

**Comment by crtrott** [comment]
Uhoh
```
/usr/bin/ld: /projects/x86-64/hpx/lib64/libhpx_wrap.a(hpx_wrap.cpp.o)(.text+0x21): unresolvable H??@?>H??FH??H??H??@?~?F?H??@?~H??8?H??H??0?FH??H??(?FH??H?? ?FH??H???FH??H???FH??H??F?fD relocation against symbol `_ZN3hpx6throwsE'
/usr/bin/ld: BFD version 2.20.51.0.2-5.47.el6_9.1 20100205 internal error, aborting at reloc.c line 443 in bfd_get_reloc_size

/usr/bin/ld: Please report this bug.

collect2: error: ld returned 1 exit status
make[2]: *** [KokkosCore_UnitTest_HostBarrier] Error 1
make[2]: *** Waiting for unfinished jobs....
/usr/bin/ld: /projects/x86-64/hpx/lib64/libhpx_wrap.a(hpx_wrap.cpp.o)(.text+0x21): unresolvable H??@?>H??FH??H??H??@?~?F?H??@?~H??8?H??H??0?FH??H??(?FH??H?? ?FH??H???FH??H???FH??H??F?fD relocation against symbol `_ZN3hpx6throwsE'
/usr/bin/ld: BFD version 2.20.51.0.2-5.47.el6_9.1 20100205 internal error, aborting at reloc.c line 443 in bfd_get_reloc_size

/usr/bin/ld: Please report this bug.

collect2: error: ld returned 1 exit status
make[2]: *** [KokkosCore_UnitTest_HWLOC] Error 1
/usr/bin/ld: BFD version 2.20.51.0.2-5.47.el6_9.1 20100205 internal error, aborting at reloc.c line 443 in bfd_get_reloc_size

/usr/bin/ld: Please report this bug.

collect2: error: ld returned 1 exit status
make[2]: *** [KokkosCore_UnitTest_HPXInterOp] Error 1
/usr/bin/ld: BFD version 2.20.51.0.2-5.47.el6_9.1 20100205 internal error, aborting at reloc.c line 443 in bfd_get_reloc_size

/usr/bin/ld: Please report this bug.

collect2: error: ld returned 1 exit status
make[2]: *** [KokkosCore_UnitTest_HPX] Error 1
make[2]: Leaving directory `/home/crtrott/Kokkos/KokkosBuild/TestAll_2019-04-06_23.46.22/hpx/1.2.1/HPX-release/core/unit_test'
make[1]: *** [all] Error 2
make[1]: Leaving directory `/home/crtrott/Kokkos/KokkosBuild/TestAll_2019-04-06_23.46.22/hpx/1.2.1/HPX-release/core/unit_test'
make: *** [build-test] Error 2
```

**Comment by crtrott** [comment]
So close ....
```
  PASSED cuda-9.1-Cuda_OpenMP-release
#######################################################
PASSED TESTS
#######################################################
clang-3.9.0-Pthread_Serial-release build_time=139 run_time=515
clang-6.0-Cuda_Pthread-release build_time=239 run_time=337
clang-6.0-OpenMP-release build_time=125 run_time=106
cuda-9.1-Cuda_OpenMP-release build_time=307 run_time=197
gcc-4.8.4-OpenMP-release build_time=142 run_time=124
gcc-4.8.4-Pthread-release build_time=137 run_time=423
gcc-5.3.0-Serial-release build_time=151 run_time=213
intel-16.0.1-OpenMP-release build_time=388 run_time=161
#######################################################
FAILED TESTS
#######################################################
hpx-1.2.1-HPX-release (build failed)
```

Looks like it is a binutils issue.

**Comment by crtrott** [comment]
Only changed the test modules for HPX, rerunning spot-check just for HPX.

**Comment by crtrott** [comment]
Ok running via spot-check ran into a weird issue where libkokkos was not done linking by the time the make system tried to use it. Other than that it works now. I am calling it and merging this in.

**Comment by msimberg** [comment]
@crtrott thanks for merging and all the help with this!

Was the "binutils issue" in the end a Makefile issue or something different? Just want to make sure this isn't an HPX problem.
